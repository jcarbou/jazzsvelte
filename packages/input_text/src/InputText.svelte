<script lang="ts">
    import type { JazzSvelteContext, HTMLInputAttributes, CssStyle, PassThroughOptions } from '@jazzsvelte/api'
    import type { TooltipOptions } from '@jazzsvelte/tooltip'
    import type { InputTextEvent, InputTextPassThroughMethodOptions, InputTextPassThroughOptions } from './inputText.types'
    import type { KeyFilterRegExp, ValidatedInputEvent } from '@jazzsvelte/key_filter'

    import { resolvePT } from '@jazzsvelte/api'
    import { createEventDispatcher, getContext } from 'svelte'
    import { tooltip, TooltipTargetDisabled } from '@jazzsvelte/tooltip'
    import { keyFilter } from '@jazzsvelte/key_filter'
    import { defaultInputTextProps as DEFAULT, globalInputTextPT as globalPt } from './inputText.config'

    export let disabled: boolean = DEFAULT.disabled
    let keyFilterType: KeyFilterRegExp | null = DEFAULT.keyFilter
    export { keyFilterType as keyFilter }
    let tooltipContent: string | null = DEFAULT.tooltip
    export { tooltipContent as tooltip }
    export let tooltipOptions: TooltipOptions | null = DEFAULT.tooltipOptions
    export let validateOnly: boolean = DEFAULT.validateOnly
    export let value: string = DEFAULT.value
    export let pt: InputTextPassThroughOptions | null = null
    export let ptOptions: PassThroughOptions | null = null
    export let unstyled: boolean = false
    let className: string | null = null
    export { className as class }
    export let style: CssStyle = null

    // TODO
    export let invalid: boolean = DEFAULT.invalid
    export let size: string | number | null = DEFAULT.size

    export const displayName = 'InptText'

    $: ptContext = {
        props: $$props,
        context: { disabled },
        ptOptions,
        unstyled
    } satisfies InputTextPassThroughMethodOptions & {
        ptOptions: PassThroughOptions | null
        unstyled: boolean
    }

    // "root element"
    $: rootAttributes = resolvePT(
        {
            class: [
                'p-inputtext p-component',
                className,
                {
                    'p-disabled': disabled,
                    'p-filled': !!value
                }
            ],
            style,
            'data-pc-name': 'inputtext',
            'data-pc-section': 'root'
        },
        pt?.root,
        globalPt?.root,
        ptContext
    ) satisfies HTMLInputAttributes

    $: showOnDisabled = !!tooltipOptions?.showOnDisabled satisfies boolean

    let jazzSvelteContext = getContext<JazzSvelteContext>('JAZZ_SVELTE')

    // Forward validatedinput event
    const dispatchValidatedInput = createEventDispatcher<InputTextEvent>()
    const onValidatedInput = (e: ValidatedInputEvent) => dispatchValidatedInput('validatedinput', e)
</script>

<TooltipTargetDisabled {showOnDisabled} useTooltip={{ tooltipContent, tooltipOptions, jazzSvelteContext }}>
    <input
        {disabled}
        {...rootAttributes}
        {...$$restProps}
        bind:value
        on:paste
        on:keydown
        on:input
        on:validatedinput={onValidatedInput}
        on:beforeinput
        use:tooltip={{ showOnDisabled, tooltipContent, tooltipOptions, jazzSvelteContext }}
        use:keyFilter={{ keyFilterType, validateOnly }}
    />
</TooltipTargetDisabled>
