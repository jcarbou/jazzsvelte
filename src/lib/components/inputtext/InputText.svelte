<script lang="ts">
    import type { HTMLInputAttributes, CssStyle } from '../utils/utils.types'
    import type { TooltipOptions } from '../tooltip/tooltipOptions.types'
    import type { InputTextEvent, InputTextPassThroughMethodOptions, InputTextPassThroughOptions } from './inputtext.types'
    import type { KeyFilterRegExp, ValidatedInputEvent } from '../keyfilter/keyfilter.types'

    import JAZZ_SVELTE from '../api/JazzSvelte'
    import { resolvePT } from '../utils/ptUtils'

    import { createEventDispatcher } from 'svelte'
    import { tooltip } from '../tooltip/tooltip.actions'
    import TooltipTargetDisabled from '../tooltip/TooltipTargetDisabled.svelte'
    import { keyFilter } from '../keyfilter/keyFilter.actions'

    export let disabled: boolean = false
    export let invalid: boolean = false
    let keyFilterType: KeyFilterRegExp | null = null
    export { keyFilterType as keyFilter }
    let tooltipContent: string | null = null
    export { tooltipContent as tooltip }
    export let tooltipOptions: TooltipOptions | undefined = undefined
    export let validateOnly: boolean = false

    export let value: string = ''
    // onInput?(event: React.FormEvent<HTMLInputElement>, validatePattern: boolean): void;

    export let pt: InputTextPassThroughOptions | null = null
    export let ptOptions: InputTextPassThroughMethodOptions | null = null
    export let unstyled: boolean = false
    let className: string | null = null
    export { className as class }
    export let style: CssStyle = null

    //$: defaultAriaLabel = label ? label + (badge ? ' ' + badge : '') : ''

    // "root element"
    $: rootAttributes = resolvePT(
        {
            class: [
                'p-inputtext p-component',
                className,
                {
                    'p-disabled': disabled,
                    'p-filled': !!value
                }
            ],
            style,
            'data-pc-name': 'inputtext',
            'data-pc-section': 'root'
        },
        pt?.root,
        JAZZ_SVELTE.pt?.inputtext?.root,
        ptOptions,
        unstyled
    ) satisfies HTMLInputAttributes

    $: showOnDisabled = !!tooltipOptions?.showOnDisabled satisfies boolean

    // Forward validatedinput event
    const dispatchValidatedInput = createEventDispatcher<InputTextEvent>()
    const onValidatedInput = (e: ValidatedInputEvent) => dispatchValidatedInput('validatedinput', e)
</script>

<TooltipTargetDisabled {showOnDisabled} useTooltip={{ content: tooltipContent, options: tooltipOptions }}>
    <input
        {disabled}
        {...rootAttributes}
        {...$$restProps}
        bind:value
        on:paste
        on:keydown
        on:input
        on:validatedinput={onValidatedInput}
        on:beforeinput
        use:tooltip={{ content: showOnDisabled ? null : tooltipContent, options: tooltipOptions }}
        use:keyFilter={{ keyFilterType, validateOnly }}
    />
</TooltipTargetDisabled>
