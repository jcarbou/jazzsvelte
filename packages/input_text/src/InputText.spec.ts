import '@testing-library/jest-dom'
import { describe, expect, vi } from 'vitest'
import { fireEvent } from '@testing-library/svelte'
import userEvent from '@testing-library/user-event'
import InputText from './InputText.svelte'
import { containerTest, later } from '@jazzsvelte/test'

describe('InputText', () => {
    containerTest('when input is enabled then input accepts data entry and have filled state', InputText, {}, async (el) => {
        const inputEl = el as HTMLInputElement
        expect(inputEl).not.toBeNull()

        if (inputEl) {
            // Act
            await fireEvent.keyDown(inputEl, { target: { value: 'abc' } })

            // Act
            expect(inputEl).toBeEnabled()
            expect(inputEl).toHaveValue('abc')
            expect(inputEl).toMatchSnapshot()
        }
    })
    containerTest('when input is blank it should not have filled state', InputText, { disabled: false }, async (inputEl) => {
        expect(inputEl).not.toBeNull()

        if (inputEl) {
            await fireEvent.input(inputEl, { target: { value: '' } })

            // Assert
            expect(inputEl).toBeEnabled()
            expect(inputEl).toHaveValue('')
            expect(inputEl).toMatchSnapshot()
        }
    })
    containerTest('when input has className only 1 className is printed', InputText, { class: 'jest' }, async (inputEl) => {
        expect(inputEl).not.toBeNull()

        if (inputEl) {
            await fireEvent.input(inputEl, { target: { value: 'jest' } })

            // Assert
            expect(inputEl).toMatchSnapshot()
            expect(inputEl).toHaveValue('jest')
        }
    })
    containerTest(
        'when input is is set for validation only',
        InputText,
        { validateOnly: true, keyFilter: `alpha` },
        async (inputEl) => {
            expect(inputEl).not.toBeNull()

            if (inputEl) {
                await fireEvent.input(inputEl, { target: { value: 'def' } })

                // Assert
                expect(inputEl).toMatchSnapshot()
                expect(inputEl).toHaveValue('def')
            }
        }
    )
    containerTest('when input is disabled it should ignore keyboard events', InputText, { disabled: true }, async (inputEl) => {
        expect(inputEl).not.toBeNull()

        if (inputEl) {
            await userEvent.type(inputEl, 'abc')

            // Assert
            expect(inputEl).toBeDisabled()
            expect(inputEl).toHaveValue('')
        }
    })
    containerTest(
        'when input is using keyFilter for integers accept integer input',
        InputText,
        { keyFilter: 'int' },
        async (el, input) => {
            const inputEl = el as HTMLInputElement | null
            expect(inputEl).not.toBeNull()
            expect(input).not.toBeNull()

            if (inputEl && input) {
                const keydownOn = vi.fn()
                input.$on('keydown', keydownOn)
                inputEl.focus()
                await userEvent.type(inputEl, '123')

                // Assert
                expect(inputEl).toHaveValue('123')
                expect(keydownOn).toHaveBeenCalledTimes(3)
            }
        }
    )
    containerTest(
        'when input is using keyFilter for integers do not accept alphabetic input',
        InputText,
        { keyFilter: 'int' },
        async (inputEl, input) => {
            expect(inputEl).not.toBeNull()
            expect(input).not.toBeNull()

            if (inputEl && input) {
                const keydownOn = vi.fn()
                input.$on('keydown', keydownOn)
                await userEvent.type(inputEl, 'abc')

                // Assert
                expect(inputEl).toHaveValue('')
                expect(keydownOn).toHaveBeenCalledTimes(3)
            }
        }
    )
    containerTest(
        'when input is using keyFilter for alphabetic accept paste of alphabetic values',
        InputText,
        { keyFilter: 'alpha' },
        async (el, input) => {
            const inputEl = el as HTMLInputElement | null
            expect(inputEl).not.toBeNull()
            expect(input).not.toBeNull()

            if (inputEl && input) {
                const pasteOn = vi.fn()
                input.$on('paste', pasteOn)
                inputEl.focus()
                await userEvent.paste('abc')

                // Assert
                expect(inputEl).toHaveValue('abc')
                expect(pasteOn).toHaveBeenCalledTimes(1)
            }
        }
    )
    containerTest(
        'when input is using keyFilter for alphabetic do not accept paste of integer values',
        InputText,
        { keyFilter: 'alpha' },
        async (el, input) => {
            const inputEl = el as HTMLInputElement | null
            expect(inputEl).not.toBeNull()
            expect(input).not.toBeNull()

            if (inputEl && input) {
                const pasteOn = vi.fn()
                input.$on('paste', pasteOn)
                inputEl.focus()
                await userEvent.paste('123')

                // Assert
                expect(inputEl).toHaveValue('')
                expect(pasteOn).toHaveBeenCalledTimes(1)
            }
        }
    )
    containerTest('when input is using onInput make sure it is called', InputText, {}, async (el, input) => {
        const inputEl = el as HTMLInputElement | null
        expect(inputEl).not.toBeNull()
        expect(input).not.toBeNull()

        if (inputEl && input) {
            const inputOn = vi.fn()
            input.$on('input', inputOn)
            inputEl.value = 'Jest testing value'
            inputEl.focus()
            await userEvent.paste(' abc')

            // Assert
            expect(inputEl).toHaveValue('Jest testing value abc')
            expect(inputOn).toHaveBeenCalledTimes(1)
        }
    })
})
