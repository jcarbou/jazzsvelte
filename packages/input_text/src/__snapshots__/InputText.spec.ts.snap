// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputText > when input has className only 1 className is printed 1`] = `
<input
  class="p-inputtext p-component jest p-filled"
  data-pc-name="inputtext"
  data-pc-section="root"
  data-tested-element="true"
/>
`;

exports[`InputText > when input is blank it should not have filled state 1`] = `
<input
  class="p-inputtext p-component"
  data-pc-name="inputtext"
  data-pc-section="root"
  data-tested-element="true"
/>
`;

exports[`InputText > when input is enabled then input accepts data entry and have filled state 1`] = `
<input
  class="p-inputtext p-component"
  data-pc-name="inputtext"
  data-pc-section="root"
  data-tested-element="true"
/>
`;

exports[`InputText > when input is is set for validation only 1`] = `
<input
  class="p-inputtext p-component p-filled"
  data-pc-name="inputtext"
  data-pc-section="root"
  data-tested-element="true"
/>
`;
